test/exercises/diffusion/heat_2d_dist_buffers.chpl

```diff
@@ -0,0 +1,149 @@
+
+  // spawn one task for each locale
+  t.start();
+  coforall (loc, (tidX, tidY)) in zip(u.targetLocales(), LOCALE_DOM) {
+    // run initialization and computation on the task for this locale
+    on loc {
+      // initialize halo arrays
+      haloArrays[tidX, tidY][N] = new haloArray(u.localSubdomain().dim(1).expand(1));
+      haloArrays[tidX, tidY][S] = new haloArray(u.localSubdomain().dim(1).expand(1));
+      haloArrays[tidX, tidY][E] = new haloArray(u.localSubdomain().dim(0).expand(1));
+      haloArrays[tidX, tidY][W] = new haloArray(u.localSubdomain().dim(0).expand(1));
+
+      // synchronize across tasks
+      b.barrier();
+
+      // run the portion of the FD computation owned by this task
+      work(tidX, tidY);
+    }
+  }
+
+  if RunCommDiag {
+    stopCommDiagnostics();
+    printCommDiagnosticsTable();
+  }
+
+  // print final results
+  const mean = (+ reduce u) / u.size,
+        stdDev = sqrt((+ reduce (u - mean)**2) / u.size);
+  t.stop();
+  writeln(abs(0.222751 - stdDev) < 1e-6);
+  if writeTime then writeln("time: ", t.elapsed());
+}
+
+proc work(tidX: int, tidY: int) {
+  // define domains to describe the indices owned by this task
+  const localIndices = u.localSubdomain(here),
+        localIndicesBuffered = localIndices.expand(1),
+        localIndicesInner = localIndices[indicesInner];
+
+	// declare two local arrays for computation on this locale
+  var uLocal1, uLocal2: [localIndicesBuffered] real = 1;
+
+  // populate local arrays with initial conditions from global array
+  uLocal1[localIndices] = u[localIndices];
+  uLocal2 = uLocal1;
+
+  // define constants for indexing into edges of local array
+  const WW = localIndicesBuffered.dim(1).low,
+        EE = localIndicesBuffered.dim(1).high,
+        SS = localIndicesBuffered.dim(0).high,
+        NN = localIndicesBuffered.dim(0).low;
+
+  // iterate for 'nt' time steps
+  for 1..nt {
+    // store results from last iteration in neighboring task's halo buffers
+    if tidY > 0       then haloArrays[tidX, tidY-1][E].v = uLocal2[.., WW+1];
+    if tidY < tidYMax then haloArrays[tidX, tidY+1][W].v = uLocal2[.., EE-1];
+    if tidX > 0       then haloArrays[tidX-1, tidY][S].v = uLocal2[NN+1, ..];
+    if tidX < tidXMax then haloArrays[tidX+1, tidY][N].v = uLocal2[SS-1, ..];
+
+    // swap local arrays
+    b.barrier();
+    uLocal1 <=> uLocal2;
+
+    // populate edges of local array from halo buffers
+    if tidY > 0       then uLocal1[.., WW] = haloArrays[tidX, tidY][W].v;
+    if tidY < tidYMax then uLocal1[.., EE] = haloArrays[tidX, tidY][E].v;
+    if tidX > 0       then uLocal1[NN, ..] = haloArrays[tidX, tidY][N].v;
+    if tidX < tidXMax then uLocal1[SS, ..] = haloArrays[tidX, tidY][S].v;
+
+    // compute the FD kernel in parallel
+    forall (i, j) in localIndicesInner do
+      uLocal2[i, j] = uLocal1[i, j] + alpha * (
+        uLocal1[i-1, j] + uLocal1[i+1, j] +
+        uLocal1[i, j-1] + uLocal1[i, j+1] -
```

Comment:
I was wondering in one of the previous versions whether the order in which we did the four +/-1 RHS reads mattered and decided it probably didn't.  I'd suggest preserving the same order for them here as in the other versions anyway, though, just for consistency for someone comparing the versions.